prusti-interface/src/specs/typed.rs

```diff
@@ -1,68 +1,157 @@
-    pub fn empty_procedure_set() -> Self {
-        SpecificationSet::Procedure(ProcedureSpecification::empty())
+/// A map of specifications keyed by crate-local DefIds.
+#[derive(Default, Debug, Clone)]
+pub struct DefSpecificationMap {
+    pub proc_specs: HashMap<LocalDefId, SpecGraph<ProcedureSpecification>>,
+    pub loop_specs: HashMap<LocalDefId, SpecGraph<LoopSpecification>>,
+    pub extern_specs: HashMap<DefId, LocalDefId>,
+}
+
+impl DefSpecificationMap {
+    pub fn new() -> Self {
+        Self::default()
     }
 
-    pub fn is_procedure(&self) -> bool {
-        matches!(self, SpecificationSet::Procedure(_))
+    pub fn get_loop_spec(&self, def_id: &DefId) -> Option<&SpecGraph<LoopSpecification>> {
+        let id = self.map_to_spec(def_id)?;
+        self.loop_specs.get(&id)
     }
 
-    #[track_caller]
-    pub fn expect_procedure(&self) -> &ProcedureSpecification {
-        if let SpecificationSet::Procedure(spec) = self {
-            return spec;
-        }
-        unreachable!("expected Procedure: {:?}", self);
+    pub fn get_proc_spec(&self, def_id: &DefId) -> Option<&SpecGraph<ProcedureSpecification>> {
+        let id = self.map_to_spec(def_id)?;
+        self.proc_specs.get(&id)
     }
 
-    #[track_caller]
-    pub fn as_procedure(&self) -> Option<&ProcedureSpecification> {
-        if let SpecificationSet::Procedure(spec) = self {
-            return Some(spec);
+    fn map_to_spec(&self, def_id: &DefId) -> Option<LocalDefId> {
+        if let Some(spec_id) = self.extern_specs.get(def_id) {
+            Some(*spec_id)
+        } else {
+            def_id.as_local()
         }
-        None
     }
+}
 
-    #[track_caller]
-    pub fn expect_mut_procedure(&mut self) -> &mut ProcedureSpecification {
-        if let SpecificationSet::Procedure(spec) = self {
-            return spec;
+#[derive(Default, Debug, Clone)]
+pub struct SpecGraph<T> {
+    /// The base specification which has no constraints
+    pub base_spec: T,
+
+    /// Specs which are active when the corresponding [SpecConstraintKind] holds on callsite
+    pub specs_with_constraints: FxHashMap<SpecConstraintKind, T>,
+}
+
+impl<T> SpecGraph<T> {
+    pub fn new(spec: T) -> Self {
+        Self {
+            base_spec: spec,
+            specs_with_constraints: FxHashMap::default(),
         }
-        unreachable!("expected Procedure: {:?}", self);
     }
+}
 
-    #[track_caller]
-    pub fn into_procedure(self) -> ProcedureSpecification {
-        if let SpecificationSet::Procedure(spec) = self {
-            return spec;
+impl SpecGraph<ProcedureSpecification> {
+    pub fn add_precondition<'tcx>(&mut self, pre: LocalDefId, env: &Environment<'tcx>) {
+        match self.get_constraint(pre, env) {
+            None => {
+                self.base_spec.pres.push(pre);
+                // Preconditions are explicitly not copied (as opposed to postconditions)
+                // This would always violate behavioral subtyping rules
```

Comment:
Maybe explain this in more detail (with an example) in a comment above `impl SpecGraph<ProcedureSpecification>`, and mention briefly that this does not happen here and *does* happen in postconditions.

Reply:
I added a big comment to `SpecGraph`. Hope it clarifies things.